runtime/major_gc.c

```diff
@@ -1237,6 +1238,16 @@ static void cycle_all_domains_callback(caml_domain_state* domain, void* unused,
 {
   uintnat num_domains_in_stw;
 
+  /* TODO: Not clear this memprof work is really part of the "cycle"
```

Comment:
@sadiqj I'd appreciate your input on this block. As the comment describes, I had a version in which this was separated out into its own function, which I called on all participating domains, but it failed because `caml_try_run_on_all_domains()` would push domains back into mark/sweep. See [this commit](https://github.com/NickBarnes/ocaml/commit/862b3b42e694c3b39af44411795dd3467dac87a0) which fails numerous tests such as `tests/memory-model/forbidden.ml`

Reply:
I'm just trying to understand what needs to be done at `caml_memprof_after_major_gc` so I can figure out where it's best placed.

I was looking at 4.14.1 and I only see `caml_memprof_update_clean_phase` being called from the major_gc. Is that the intention for this hook, that it should be called at the point where ephemerons need cleaning? That's a different place from where it is here.

Reply:
It needs to be called after marking is complete, for a similar purpose to cleaning ephemerons - memprof needs to walk its internal data structures, to update all entries for blocks which have not survived collection. Originally I put the call in `major_collection_slice`, close to the ephemerons stuff, but that didn't work as previously noted. It's ended up here as a placeholder, really.